packages/labs/ssr/src/test/lib/render-lit_test.ts

```diff
@@ -545,6 +552,133 @@ for (const global of [emptyVmGlobal, shimmedVmGlobal]) {
+    const {render, serverOnlyRenderHydratable} = await setup();
+    const result = render(serverOnlyRenderHydratable);
+    assert.is(
+      result,
+      `
+    <div>server only</div>
+    <!--lit-part AEmR7W+R0Ak=--><div><!--lit-part-->hydratable<!--/lit-part--></div><!--/lit-part-->
+  `
+    );
+  });
+
+  test(`normal template can't render a server-only template`, async () => {
+    const {render, hydratableRenderServerOnly} = await setup();
+    assert.throws(
+      () => render(hydratableRenderServerOnly),
+      /A server-only template can't be rendered inside an ordinary, hydratable template./
+    );
+  });
+
+  test('server-only template into raw element', async () => {
+    const {render, serverOnlyRawElementTemplate} = await setup();
+    const result = await render(serverOnlyRawElementTemplate);
+    assert.is(
+      result,
+      `
+    <title>No comments inside</title>
+    <textarea>This also works (kinda).</textarea>
+  `
+    );
+  });
+
+  test('server-only template into a <template> element', async () => {
+    const {render, serverOnlyInTemplateElement} = await setup();
+    const result = await render(serverOnlyInTemplateElement);
+    assert.is(
+      result,
+      `
+    <template>one<div>two<div>three</div><template>recursed</template></div></template>
+  `
+    );
+  });
+
+  test('server-only document template can render an entire document', async () => {
+    const {render, serverOnlyDocumentTemplate} = await setup();
+    const result = await render(serverOnlyDocumentTemplate);
+    assert.is(
+      result,
+      `
+    <!DOCTYPE html>
+    <html>
+      <head>
+        <title>No comments inside</title>
+      </head>
+      <body>
+        <textarea>This also works (kinda).</textarea>
+      </body>
+    </html>
+  `
+    );
+  });
+
+  test('server-only template throws on property bindings', async () => {
+    const {render, serverOnlyRenderPropertyBinding} = await setup();
+    assert.throws(
+      () => render(serverOnlyRenderPropertyBinding),
+      /Server-only templates can't bind to properties./
+    );
+  });
+
+  test('server-only template throws on event bindings', async () => {
+    const {render, serverOnlyRenderEventBinding} = await setup();
+    assert.throws(
+      () => render(serverOnlyRenderEventBinding),
+      /Server-only templates can't bind to events./
+    );
+  });
+
+  test('server-only template into a <script>', async () => {
+    const {render, renderServerOnlyScript, renderServerOnlyScriptDeep} =
+      await setup();
+    assert.throws(
+      () => render(renderServerOnlyScript),
+      /Found binding inside an executable <script> tag in a server-only template\./
+    );
+    assert.throws(
+      () => render(renderServerOnlyScriptDeep),
+      /Found binding inside an executable <script> tag in a server-only template\./
+    );
+  });
+
+  test('server-only template into non-executing script', async () => {
```

Comment:
Worth adding an escape hatch in this PR for providing known safe executable scripts?
Also, what about `<script>` inside a `<template>` within a server template?

Reply:
IMO this PR is already too big, and that's a fairly fringe use case. Let's do that separately.

A script inside a template is safe, but harder to detect, and there aren't clear use cases that I can see.